### PR TITLE
S15 t2 update the worker availability according to the specific worker timeslot

### PR DESCRIPTION
**Pull request by Simran230**
Scrum Board Ticket: S15 T1 Update the worker availability according to the specific workerTimeslotId

Story: Story 15: Update service worker time offerings

Description: Attempts to update the time offering of a worker given a workerTimeslotId. However, if no row is actually updated the method should throw an error. 

Run the code locally and ensure no complications when starting up (run the command npm run-script start:dev in the root folder OR run the command npm start if your node and npm versions are above 12.10.0 for node and 6.10.3 for npm).

Running the automated server test codes and ensured that there was 100% file coverage (and passing tests) for all the unit tests. To do this, navigate inside the server folder and run npm test.

Open Postman and test the endpoint by sending requests to the /api/book-appointment endpoint on port 4000 with various combinations of query parameter workerTimeslotId. Cross-reference the response with your local database values to ensure that the response is as expected.
Make sure to specify an accessToken as this is a private endpoint. An example of an accessToken is XcCa92ZvOnQKZsGtOKOa

### CODE REVIEW

File: server/models/db/workerTimeslot.js
Comment (nancyprabhakar) [inline server/models/db/workerTimeslot.js]: Hi Simran,

I do not believe that you need the `workerId` and `date` parameters for this query. I would recommend finding the status corrsponding to only the workerTimeslotId parameter in the Worker_Timeslot table. You can look at the     `server/migrations/009.do.create-worker-timeslot-table.sql` for reference.

File: server/models/db/workerTimeslot.js
Comment (nancyprabhakar) [inline server/models/db/workerTimeslot.js]: Hi Simran,

This query statement updates the `status` to unavailable regardless of the initial value of the `status` parameter. I would recommend only changing the `status` to 'unavailable' if the original `status` was 'available'. If the status value is originally 'unavailable', there is no need to update the status parameter.

File: server/models/handlers/workerTimeslot.js
Comment (nancyprabhakar) [inline server/models/handlers/workerTimeslot.js]: Hi Simran,

This try-catch block checks the insertWorkerTimeslot method, which has been previously tested. I would recommend that you place the try-catch method before the `updatingTheWorkerAvailability(status)` so that this new method is tested.

File: server/models/handlers/workerTimeslot.js
Comment (nancyprabhakar) [inline server/models/handlers/workerTimeslot.js]: Hi Simran,

I'm curious: What happens if the `status` is 'null'?

File: server/test/handlers/updateAvailability.test.js
Comment (nancyprabhakar) [inline server/test/handlers/updateAvailability.test.js]: Hi Simran,

This is a version 1 of the PR, so I understand there are no tests created yet. However, to proceed with this PR, I'd recommend writing automated test cases to cover the `updatingTheWorkerAvailability` method.

**Review by nancyprabhakar — COMMENTED** [review]
Hi Simran,

Awesome work so far! I've made some comments regarding your query statement which I'd like addressed before I can approve this PR. I'd also recommend replacing the `workerId` and `date` parameters with `workerTimeslotId` for code brevity. Please let me know when you've made the requested changes and I can take another look!

Thanks :)
Nancy

**Comment by Simran230** [comment]
Had a working session with Peter for 1.5 hours. We tested by creating a dummy endpoint with 100% converge in testing and created a endpoint for the handlers. Before merging into master we're waiting for S15T1 to be completed and remove the dummy endpoint to complete this task.

**Comment by prdbondi** [comment]
Hey Simran, thank you for the great work we had on the call a second ago! 
I have successfully followed all testing steps outlined in this PR and am happy to report that testing passed at 100% coverage and I have tried different values in endpoint testing and all seems to go as expected. 
I would approve current state, but as you mentioned, there needs to be some integration testing done with S15 T1 so I will leave this PR open until we can follow up on that.

File: server/models/db/workerTimeslot.js
Comment (Simran230) [reply server/models/db/workerTimeslot.js]: Hi Nancy! 

Thank you very much for your comments! Peter and I have amended the changes to reflect your comments in the latest commits. I appreciate all your help and time! 

Simran

File: server/models/db/workerTimeslot.js
Comment (Simran230) [reply server/models/db/workerTimeslot.js]: Hi Nancy! 

Thanks for catching this! Peter and I have implemented changes in the latest PR to ensure that your comment can be resolved. Thank you very much Nancy! 

Simran

File: server/models/handlers/workerTimeslot.js
Comment (Simran230) [reply server/models/handlers/workerTimeslot.js]: Hi Nancy, 

That's a great question! I will communicate with the team regarding that question. I was assuming that the status can only be, "available" and "unavailable"? You've addressed a good point and I will research on this further! 

Thanks! 

Simran

File: server/test/handlers/updateAvailability.test.js
Comment (Simran230) [reply server/test/handlers/updateAvailability.test.js]: Hi Nancy! 

Thanks a lot! Peter and I have created 3 tests in our latest PR to resolve this comment. Thank you for your time! 

Simran

File: server/models/handlers/workerTimeslot.js
Comment (Simran230) [reply server/models/handlers/workerTimeslot.js]: Hi Nancy! 

Once again great catch! I will look into this comment and do some testing to ensure that the try-catch method is before the status. Thank you for your time! 

Simran

File: server/test/handlers/bookWorkerTimeslot.test.js
Comment (nancyprabhakar) [inline server/test/handlers/bookWorkerTimeslot.test.js:3]: Hi Simran,
I noticed that for test case 1 and 2, the status of the workerTimeslotId = 5 and 7 must be unavailable. The same database is being used by multiple methods which edit the status of the entries in the worker_Timeslot table (eg.  S19-T2 changes the status from unavailable to available when an appointment is cancelled). Due to this, the value of the status parameter is constantly changing. Since your tests return true (indicating that the slot is unavailable), these tests may not return true if the value of status in the database for workerTimeslotId = 5 and 7 is unavailable.

File: server/test/handlers/updateAvailability.test.js
Comment (nancyprabhakar) [reply server/test/handlers/updateAvailability.test.js]: Hi Simran,

I took a look at your test cases, thank you for adding those in!
I noticed that for test case 1 and 2, the status of the workerTimeslotId = 5 and 7 must be unavailable. The same database is being used by multiple methods which edit the status of the entries in the worker_Timeslot table (eg. S19-T2 changes the status from unavailable to available when an appointment is cancelled). Due to this, the value of the status parameter is constantly changing. Since your tests return true (indicating that the slot is unavailable), these tests may not return true if the value of status in the database for workerTimeslotId = 5 and 7 is unavailable.

File: server/models/handlers/workerTimeslot.js
Comment (nancyprabhakar) [reply server/models/handlers/workerTimeslot.js]: Sounds good! As long as there is 100% test coverage, you should be good to go!

File: server/models/db/workerTimeslot.js
Comment (nancyprabhakar) [reply server/models/db/workerTimeslot.js]: Awesome! I'm glad you were able to address this!

File: server/models/db/workerTimeslot.js
Comment (nancyprabhakar) [reply server/models/db/workerTimeslot.js]: Awesome! I'm glad you were able to address this!

**Review by nancyprabhakar — APPROVED** [review]
This is awesome Simran! Thank you for making the requested changes and clarifying my concerns.

As per the dependencies for this ticket, I understand that there is pending integration required with S15 T1: Determine if worker is available during worker timeslot. I also notice that pending this integration, the API being leveraged is a dummy API: api/book-appointment-test-worker-timeslot-update, which will be switched to the correct API upon integration. As such, I will be approving this PR, but I will hold off on merging this PR until the integration with S15 T1 has been complete. I welcome Peter to merge the PR (the second reviewer) as he will be assisting in the integration. 

I ran the automated test in the server folder (as there is no front end UI set up for this yet) with the command npm test.
The test passed and the code coverage is at 100%. I also manually tested the endpoint locally by running npm start in the server folder. The output matches the expected result perfectly!
Requests sent to the api/book-appointment-test-worker-timeslot-update on Postman (with the parameters: workerTimeslotId 5 and 7  and accessToken eeJAQr3wEC6CJZROFJTY) gives the response which is expected.

As such, I am approving this PR :) Thank you for addressing the concerns I had mentioned in my initial review and for implementing the requested changes.

File: server/test/handlers/updateAvailability.test.js
Comment (Simran230) [reply server/test/handlers/updateAvailability.test.js]: Hi Nancy, 

Thank you very much for bringing this point up. Peter and I had to carefully choose workerTimeslotId 5 and 7 to test since it's not being changed or altered by other tests. In this case the status will not change from available to unavailable by another test other than ours. 

Thank you very much for bringing such a critical point up! I appreciate it! 

Simran

**Review by prdbondi — APPROVED** [review]
Completed integration testing via npm test -> 100% and Postman endpoint testing on call with Simran for S15 T1 and S15 T2. Book appointment endpoint works as expected. 
See snippet of endpoint testing screenshotted. 
This PR, as nancy said, is ready to merge

**Pull request merged**
